dftimewolf/cli/dftimewolf_recipes.py

```diff
@@ -275,6 +277,27 @@ def ParseArguments(self, arguments: List[str]) -> None:
 
     self._state.command_line_options = vars(self._command_line_options)
 
+  def ValidateArguments(self) -> None:
+    """Validate the arguments."""
+    recipe = self._recipes_manager.Recipes()[self._recipe['name']]
+    error_messages = []
+
+    for arg in recipe.args:
+      try:
+        switch = arg.switch.replace('--', '')
+        self._args_validator.Validate(self.state.command_line_options[switch],
+                                      arg.format)
+      except errors.RecipeArgsValidatorError as exception:
+        error_messages.append(f'Argument validation error: "{arg.switch}" with '
+            f'value "{self.state.command_line_options[switch]}" gave error: '
+            f'{str(exception)}')
+
+    if error_messages:
+      for message in error_messages:
+        logger.critical(message)
+      raise errors.RecipeArgsValidatorError(
+          'At least one argument failed validation')
```

Comment:
I think this would be best placed directly in argparse using the `type` parameter of `add_argument`: https://docs.python.org/3/library/argparse.html#type

If your validators can be reduced to simple functions, then we can just map the string provided in the `format` key of the arg with your validation function.

Or, given that the set of parameters can be a bit complex (e.g. `regex` depends on the actual regex being passed), you could also imagine a function that takes in the last element of the tuple in `args`, and passes that to a function that "picks" which validator to run, with which parameters.

In addition to validating parameters, this would also enable us to e.g. split comma-separated values before gettingp assed to SetUp(), which would avoid us having to write this logic in every module.

WDYT?

Reply:
Overall, that makes sense, I'll give it a crack and see if I can get something working. That said:

> If your validators can be reduced to simple functions, then we can just map the string provided in the format key of the arg with your validation function.

They probably can, but, from the docs: `The argument to type can be any callable that accepts a single string.` Being a single string makes it tough, since we can't get the optional parameters in that way. Even with your suggestion `you could also imagine a function that takes in the last element of the tuple in args`, we need to pass in a dict (or convert it to, and then back from, a `str`.)

> In addition to validating parameters, this would also enable us to e.g. split comma-separated values before gettingp assed to SetUp()

This would be ideal, but I suspect it might end up as a breaking change, which I was hoping to avoid - As it will mean that all modules need to be updated to accept a list where before they required a single item.

Reply:
I've put together a comparison PR, #706 that starts down the route of using validator functions in the `type` parameter of `argparse.add_parameter`. There's a few drawbacks, including less control over error messages, losing the comma separation, and some pretty funky function signatures which IMO impacts readability + maintainablility. We'd also need a new validator instance for each validation, since now the params are stored as class members (so, we can't reuse a single Regex Validator for multiple regexes in the same recipe, for example.)

Reply:
Yup, that's fair, thanks for taking time to write the alternate PR. It's much clearer the way it's done here.